website/server/libs/googlePayments.js

```diff
@@ -0,0 +1,156 @@
+    _id: token,
+  }).exec();
+  if (existingReceipt) throw new NotAuthorized('RECEIPT_ALREADY_USED');
+
+  await IapPurchaseReceipt.create({
+    _id: token,
+    consumed: true,
+    userId: user._id,
+  });
+
+  let amount;
+
+  switch (receiptObj.productId) {
+    case 'com.habitrpg.android.habitica.iap.4gems':
+      amount = 1;
+      break;
+    case 'com.habitrpg.android.habitica.iap.20.gems':
+    case 'com.habitrpg.android.habitica.iap.21gems':
+      amount = 5.25;
+      break;
+    case 'com.habitrpg.android.habitica.iap.42gems':
+      amount = 10.5;
+      break;
+    case 'com.habitrpg.android.habitica.iap.84gems':
+      amount = 21;
+      break;
+  }
+
+  if (!amount) throw new NotAuthorized('INVALID_ITEM_PURCHASED');
+
+  await payments.buyGems({
+    user,
+    paymentMethod: 'IAP GooglePlay',
+    amount,
+    headers,
+  });
+
+  return googleRes;
+};
+
+api.subscribe = async function subscribe (sku, user, receipt, signature, headers, nextPaymentProcessing = undefined) {
+  if (!sku) throw new BadRequest(shared.i18n.t('missingSubscriptionCode'));
+  let subCode;
+  switch (sku) {
+    case 'com.habitrpg.android.habitica.subscription.1month':
+      subCode = 'basic_earned';
+      break;
+    case 'com.habitrpg.android.habitica.subscription.3month':
+      subCode = 'basic_3mo';
+      break;
+    case 'com.habitrpg.android.habitica.subscription.6month':
+      subCode = 'basic_6mo';
+      break;
+    case 'com.habitrpg.android.habitica.subscription.12month':
+      subCode = 'basic_12mo';
+      break;
+  }
+  let sub = subCode ? shared.content.subscriptionBlocks[subCode] : false;
+  if (!sub) throw new NotAuthorized('INVALID_ITEM_PURCHASED');
+
+  await iap.setup();
+
+  let testObj = {
+    data: receipt,
+    signature,
+  };
+
+  let receiptObj = JSON.parse(receipt); // passed as a string
+  let token = receiptObj.token || receiptObj.purchaseToken;
+
+  let existingUser = await User.findOne({
+    'payments.plan.customerId': token,
+  }).exec();
+  if (existingUser) throw new NotAuthorized('RECEIPT_ALREADY_USED');
+
+  let googleRes = await iap.validate(iap.GOOGLE, testObj);
+
+  let isValidated = iap.isValidated(googleRes);
+  if (!isValidated) throw new NotAuthorized('INVALID_RECEIPT');
+
+  nextPaymentProcessing = nextPaymentProcessing ? nextPaymentProcessing : moment.utc().add({days: 2});
+
+  await payments.createSubscription({
+    user,
+    customerId: token,
+    paymentMethod: 'Google',
+    sub,
+    headers,
+    nextPaymentProcessing,
+    additionalData: testObj,
+  });
+};
+
+
+api.cancelSubscribe = async function cancelSubscribe (user, headers) {
+  let data = user.purchased.plan.additionalData;
+
+  if (!data) throw new NotAuthorized(shared.i18n.t('missingSubscription'));
+
+  await iap.setup();
+
+  let googleRes = await iap.validate(iap.GOOGLE, data);
+
+  let isValidated = iap.isValidated(googleRes);
+  if (!isValidated) throw new NotAuthorized('INVALID_RECEIPT');
```

Comment:
The app uses this value to determine how to handle the response. so changing it to be translated would break that. Also it's not shown to the user anywhere and these api calls shouldn't be used outside of our apps.